### PR TITLE
extlinux-conf-builder: expose and use base builder command, allow a custom FDT to be specified

### PR DESCRIPTION
**Pull request by flokli**
Some bootloaders might not properly detect the model.
If the specific model is known by configuration, provide a way to
explicitly point to a specific dtb in the extlinux.conf.

This adds `hardware.deviceTree.name`, uses it from the generic-extlinux-compatible module, and exposes the builder command generated there as a read-only NixOS option, which is now used by the sd-image-*.nix files - allowing us to get rid of all the additional imports there, too.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc @NixOS/exotic-platform-maintainers

### CODE REVIEW

File: nixos/modules/system/boot/loader/generic-extlinux-compatible/extlinux-conf-builder.sh [hunk 1]
```diff
@@ -95,7 +96,9 @@ addEntry() {
     fi
     echo "  LINUX ../nixos/$(basename $kernel)"
     echo "  INITRD ../nixos/$(basename $initrd)"
-    if [ -d "$dtbDir" ]; then
+    if [ -n "$dtbName" ]; then
+        echo "  FDT ../nixos/$(basename $dtbs)/${dtbName}.dtb"
+    elif [ -d "$dtbDir" ]; then
         echo "  FDTDIR ../nixos/$(basename $dtbs)"
     fi
```
Comment (samueldr) [inline nixos/modules/system/boot/loader/generic-extlinux-compatible/extlinux-conf-builder.sh:103]: The logic here seems a bit screwy. I think the script's `-n` option should fail if `dtbDir` ends up not set.

Comment (flokli) [reply nixos/modules/system/boot/loader/generic-extlinux-compatible/extlinux-conf-builder.sh:103]: right - but once `-n` is set, we'd still need to pick up from `dtbDir`.

Comment (flokli) [reply nixos/modules/system/boot/loader/generic-extlinux-compatible/extlinux-conf-builder.sh]: pushed a new version, PTAL.

File: nixos/modules/system/boot/loader/generic-extlinux-compatible/extlinux-conf-builder.sh [hunk 2]
```diff
@@ -6,7 +6,7 @@ export PATH=/empty
 for i in @path@; do PATH=$PATH:$i/bin; done
 
 usage() {
-    echo "usage: $0 -t <timeout> -c <path-to-default-configuration> [-d <boot-dir>] [-g <num-generations>]" >&2
+    echo "usage: $0 -t <timeout> -c <path-to-default-configuration> [-d <boot-dir>] [-g <num-generations>] [-d <dtbName>]" >&2
```
Comment (samueldr) [inline nixos/modules/system/boot/loader/generic-extlinux-compatible/extlinux-conf-builder.sh:9]:
```suggestion
    echo "usage: $0 -t <timeout> -c <path-to-default-configuration> [-d <boot-dir>] [-g <num-generations>] [-n <dtbName>]" >&2
```

File: nixos/modules/system/boot/loader/generic-extlinux-compatible/extlinux-conf-builder.sh [hunk 3]
```diff
@@ -96,7 +97,12 @@ addEntry() {
     echo "  LINUX ../nixos/$(basename $kernel)"
     echo "  INITRD ../nixos/$(basename $initrd)"
     if [ -d "$dtbDir" ]; then
-        echo "  FDTDIR ../nixos/$(basename $dtbs)"
+        # if a dtbName was specified explicitly, use that, else use FDTDIR
+        if [ -n "$dtbName" ]; then
+            echo "  FDT ../nixos/$(basename $dtbs)/${dtbName}.dtb"
+        else
+            echo "  FDTDIR ../nixos/$(basename $dtbs)"
+        fi
```
Comment (samueldr) [inline nixos/modules/system/boot/loader/generic-extlinux-compatible/extlinux-conf-builder.sh:105]: This is a bit better than the previous, but has the fatal flaw of not failing if `-n` is given, but `$dtbDir` isn't a thing.

I guess in a new `else` branch, fail on `[ -n "$dtbName" ]` wasn't given.

File: nixos/modules/system/boot/loader/generic-extlinux-compatible/extlinux-conf-builder.sh (hunk 2 above)
Comment (flokli) [reply nixos/modules/system/boot/loader/generic-extlinux-compatible/extlinux-conf-builder.sh]: woops, fixed!

File: nixos/modules/system/boot/loader/generic-extlinux-compatible/extlinux-conf-builder.sh (hunk 3 above)
Comment (flokli) [reply nixos/modules/system/boot/loader/generic-extlinux-compatible/extlinux-conf-builder.sh:105]: well, it also doesn't check for the fdt to be really present, and even with a specific fdt specified, still copies all of them.

Not sure how much logic we actually want to add in here…

Comment (flokli) [reply nixos/modules/system/boot/loader/generic-extlinux-compatible/extlinux-conf-builder.sh:105]: I added the new else branch as described.

**Review by samueldr — APPROVED** [review]
I haven't run the code, but the author verifiably did, successfully, and it looks all right.

**Comment by flokli** [comment]
I realized this currently only fixes half of the problem - while one would now be able to pass this to the exlinux builder, the `generic-extlinux-compatible` is in charge of invoking the builder on the next configuration switch - and it doesn't know about the custom fdt specified, so it'd remove that option again, which is most likely not what we want.

I added the necessary option as `hardware.deviceTree.name`, use it from the `generic-extlinux-compatible` module, and expose the builder command generated there as a read-only NixOS option, which is now used by the `sd-image-*.nix` files - allowing us to get rid of all the additional imports there, too.

**Comment by flokli** [comment]
I pushed an updated version, containing some drive-by fixes, the refactor, and the addition of the fdt option, please take another look :-)

File: nixos/modules/system/boot/loader/generic-extlinux-compatible/default.nix [hunk 4]
```diff
@@ -8,7 +8,7 @@ let
 
   timeoutStr = if blCfg.timeout == null then "-1" else toString blCfg.timeout;
 
-  builder = import ./extlinux-conf-builder.nix { inherit pkgs; };
+  builder = import ./extlinux-conf-builder.nix { pkgs = pkgs.buildPackages; };
```
Comment (flokli) [inline nixos/modules/system/boot/loader/generic-extlinux-compatible/default.nix:12]: This is still not correct. While this now succeeds in assembling the image, switching at runtime won't work.

I feel like the one we expose for image generation should set  `pkgs = pkgs.buildPackages`) (and should probably be named to reflect that), but we still need to have a builderusing the native packages inside this module (as that generates the line in the activation script, which runs on the board architecture).

Comment (flokli) [reply nixos/modules/system/boot/loader/generic-extlinux-compatible/default.nix]: I renamed the option to `populateCmd`, so it's usecase is clearer, and explicitly distinguished between the builder used during activation and the builder used to populate a (possibly cross) image.

I prefer this over just exposing the arguments - then every `sd-image-*` would then need to still import the builder.

**Comment by flokli** [comment]
This now should also work - I successfully switched configurations and built cross sd images with this module. PTAL.

**Review by samueldr — APPROVED** [review]
I haven't run the latest changes, but it all looks fine and I trust @flokli as having ran this correctly.

The newly added option (`hardware.deviceTree.name`) may be useful in other situations too.

**Comment by Ericson2314** [comment]
I....totally don't know about this stuff. But maybe the other @NixOS/exotic-platform-maintainers do and can review instead (I see you already CCed).

**Comment by flokli** [comment]
Okay, let's merge this :-)

**Pull request merged**